project_material/models/project_task_material.py

```diff
@@ -0,0 +1,317 @@
+    unit_cost = fields.Float(
+        related='product_id.standard_price',
+        string='Unit Cost',
+        readonly=True,
+    )
+    move_ids = fields.One2many(
+        'stock.move',
+        'material_line_id',
+        'Stock Moves',
+    )
+
+    @api.depends(
+        'move_ids.move_orig_ids',
+        'move_ids.move_orig_ids.state',
+        'move_ids.move_orig_ids.returned_move_ids',
+        'move_ids.move_orig_ids.returned_move_ids.state',
+    )
+    def _compute_prepared_qty(self):
+        for line in self:
+            preparation_moves = line.mapped('move_ids.move_orig_ids')
+            preparation_moves_done = preparation_moves.filtered(lambda m: m.state == 'done')
+            prepared_qty = sum(preparation_moves_done.mapped('product_uom_qty'))
+            return_moves = preparation_moves_done.mapped('returned_move_ids')
+            return_moves_done = return_moves.filtered(lambda m: m.state == 'done')
+            returned_qty = sum(return_moves_done.mapped('product_uom_qty'))
+            line.prepared_qty = prepared_qty - returned_qty
+
+    @api.depends(
+        'move_ids',
+        'move_ids.state',
+    )
+    def _compute_consumed_qty(self):
+        for line in self:
+            consumed_moves = line.move_ids.filtered(
+                lambda m: m.state == 'done' and m.picking_code == 'consumption')
+            consumed_qty = sum(consumed_moves.mapped('product_uom_qty'))
+            return_moves = line.move_ids.filtered(
+                lambda m: m.state == 'done' and m.picking_code == 'consumption_return')
+            returned_qty = sum(return_moves.mapped('product_uom_qty'))
+            line.consumed_qty = consumed_qty - returned_qty
+
+    @api.model
+    def create(self, vals):
+        """Generate procurements when adding a new material line.
+
+        The sudo is required, because project users do not have access to stock objects.
+        """
+        line = super().create(vals)
+        if line._should_generate_procurement():
+            line.sudo()._run_procurements()
+        return line
+
+    @api.multi
+    def write(self, vals):
+        """Adjust procurements when modifying the quantity on a material line.
+
+        The sudo is required, because project users do not have access to stock objects.
+        """
+        if 'product_id' in vals or 'task_id' in vals:
+            for line in self:
+                line.sudo()._check_can_change_product_or_task()
+                line.sudo()._cancel_procurements()
+
+        super().write(vals)
+
+        if 'product_id' in vals or 'task_id' in vals or 'initial_qty' in vals:
+            lines_with_procurement = self.filtered(lambda l: l._should_generate_procurement())
+            for line in lines_with_procurement:
+                line.sudo()._run_procurements()
+
+        return True
+
+    @api.multi
+    def unlink(self):
+        """Cancel procurements when deleting material line.
```

Comment:
Codacy found an issue: [Multi-line docstring summary should start at the second line (D213)](https://app.codacy.com/gh/Numigi/odoo-project-addons/pullRequest?prid=6165869)